### PR TITLE
Feature/28 track

### LINKED ISSUE #28: Implement interface for a `track`
## User stories
As a developer I want to have access to the tracks in order to manipulate and query them

## Dependencies
Depends on:
 - #5
 - #15

## (Wiki) References
- #10 

## Acceptance Criteria
- [ ] ~~get blocksections per Fahrstrasse~~
- [x] get max-speed
- [x] get blocked-state of the edge
- [ ] ~~update the train to return a track_segment when queried for a track~~ Will be fixed in #215

### PR DESCRIPTION
**Pull request by SaturnHafen**
Fixes #28

## PR checklist

- [x] Acceptance criteria fulfilled
- [x] Additional features are tested
- [x] Docs (code, wiki & diagrams) updated
- [ ] ~~Breaking changes anounced~~
- [x] Dev-branch has been merged into local branch to resolve conflicts
- [x] Tests and linter have passed AFTER local merge
- [x] Another dev reviewed and approved

### CODE REVIEW

File: src/wrapper/simulation_objects.py [hunk 1]
```diff
@@ -84,3 +84,38 @@ def update(self, data: dict):
...
+        """Updates the max_speed of the edge
+        performance consideration: This method performs a traci call
+
+        :param max_speed: The new maximum speed of the edge
+        """
+        edge.setMaxSpeed(self.identifier, max_speed)
+        self._max_speed = max_speed
+
+    def __init__(self, identifier: str = None):
+        SimulationObject.__init__(self, identifier)
```
Comment (CR1337) [inline src/wrapper/simulation_objects.py]: `super().__init__` 

See above.

File: src/wrapper/simulation_object_updating_component.py [hunk 2]
```diff
@@ -11,8 +14,40 @@ class SimulationObjectUpdatingComponent(Component):
...
+        :return: The tracks in the simulation
+        """
+        return [
+            x
+            for x in self._simulation_objects
+            if isinstance(x, src.wrapper.simulation_objects.Track)
+        ]
+
+    def __init__(self, logger=None):
+        Component.__init__(self, priority=10, logger=logger)
```
Comment (CR1337) [inline src/wrapper/simulation_object_updating_component.py]: Probably it doesn't make a difference here but the cleaner solution would be to call `super().__init__`. 
See this Post: https://stackoverflow.com/questions/42413670/whats-the-difference-between-super-and-parent-class-name

super calls in python are tricky 😬

**Review by CR1337 — DISMISSED** [review]
I don't know entirely what's happening here but it look reasonable.

Just a minor suggestions regarding super calls.

File: src/wrapper/simulation_object_updating_component.py [hunk 3]
```diff
@@ -14,6 +14,26 @@ class SimulationObjectUpdatingComponent(Component):
 
     _simulation_objects = None
 
+    @property
+    def simulation_objects(self) -> List["SimulationObject"]:
+        """Returns a list of all objects in the simulation
```
Comment (instant-sky) [inline src/wrapper/simulation_object_updating_component.py]: You could import `SimulationObject` from `src.wrapper.simulation_objects` or directly use `src.wrapper.simulation_objects.SimulationObject`. This then makes it possible to omit the "" and enables Pylance to recognize these classes.

File: src/wrapper/simulation_object_updating_component.py [hunk 4]
```diff
@@ -14,6 +14,26 @@ class SimulationObjectUpdatingComponent(Component):
...
+
+        :return: the objects in the simulation
+        """
+        return self._simulation_objects
+
+    @property
+    def signals(self) -> List["Signal"]:
```
Comment (instant-sky) [inline src/wrapper/simulation_object_updating_component.py]: Same as in the last comment.

File: src/wrapper/simulation_object_updating_component.py [hunk 5]
```diff
@@ -14,6 +14,26 @@ class SimulationObjectUpdatingComponent(Component):
...
+        :return: The signals in the simulation
+        """
+        return [
+            x
+            for x in self._simulation_objects
+            if isinstance(x, src.wrapper.simulation_objects.Signal)
+        ]
+
     @property
     def platforms(self) -> List["Platform"]:
```
Comment (instant-sky) [inline src/wrapper/simulation_object_updating_component.py]: see above

File: src/wrapper/simulation_object_updating_component.py [hunk 6]
```diff
@@ -26,8 +46,20 @@ def platforms(self) -> List["Platform"]:
             if isinstance(x, src.wrapper.simulation_objects.Platform)
         ]
 
-    def __init__(self, *args, **kwargs):
-        Component.__init__(self, *args, **kwargs)
+    @property
+    def tracks(self) -> List["Track"]:
```
Comment (instant-sky) [inline src/wrapper/simulation_object_updating_component.py]: see above

**Review by instant-sky — COMMENTED** [review]
Nice feature! I am excited to use the tracks in the fault injector. 🛤️

File: tests/wrapper/test_simulation_objects.py [hunk 7]
```diff
@@ -30,6 +30,36 @@ def test_update_state(self, traffic_update):
...
+    def speed_update(self, monkeypatch):
+        def set_max_speed(identifier: str, speed: float) -> None:
+            assert identifier == "track"
+            assert speed == 100
+
+        monkeypatch.setattr(edge, "setMaxSpeed", set_max_speed)
+
+    def test_update_speed(self, track, speed_update):
+        # pylint: disable=unused-argument
+        track.max_speed = 100
```
Comment (instant-sky) [inline tests/wrapper/test_simulation_objects.py:50]: Shouldn't the setter method be used here?

File: src/wrapper/simulation_object_updating_component.py (hunk 3 above)
Comment (SaturnHafen) [reply src/wrapper/simulation_object_updating_component.py]: The problem with that is then you get circular-import failures. That is why I didn't import them

File: tests/wrapper/test_simulation_objects.py (hunk 7 above)
Comment (SaturnHafen) [reply tests/wrapper/test_simulation_objects.py:50]: It should use the setter automatically, thanks to [this](https://github.com/BP2022-AP1/bp2022-ap1/pull/210/files/e20e8d4091060d5b77701fab7fcb912f1d537129#diff-32b59e5349657358f01749afb728eedcd06c4ca2ca854f0b475cc29a31aa8ef5R119) annotation

Comment (instant-sky) [reply tests/wrapper/test_simulation_objects.py:50]: Ah cool, I didn't know that. 🤯

**Review by instant-sky — APPROVED** [review]

**Pull request merged**